blog-posts/angular-inject-the-injector.md

```diff
@@ -0,0 +1,252 @@
+  }
+}
+```
+
+Can you see the elegance here? We inject the injector, which is the engine of the dependency injection (DI) system in Angular itself and then request the singleton instances of our desired services to assign them to the local variables as before.
+
+The injector has the correct typing, that the result might be undefined, that's why we convince TypeScript with the [non-null assertion operator (!)](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator) that we guarantee to receive a value. That's possible because we control the environment, and the services are guaranteed to be available as they are provided in the root (`@Injectable({providedIn: 'root' })`).
+
+In case we decide to add a fourth of fifth service in the base class, we can now add it and request it directly from the injector without breaking any subclass as the constructor signature stays the same. See how many lines of random services the team are not even interested in, can now be replaced:
+
+```diff-typescript
+@Component({
+  selector: 'my-unsubscribe-tool',
+  template: `<button (click)="unsubscribe()">unsubscribe</button>`
+})
+export class UnsubscribeToolComponent extends MyAbstractBaseComponent {
+  constructor(
+    @Inject(Injector) injector: Injector
+-   @Inject(MySubscribeService) mySubscribeService: MySubscribeService,
+-   @Inject(MyTrackingService)  myTrackingService: MyTrackingService,
+-   @Inject(MyRemoteHomeService) public  myRemoteHomeService: MyRemoteHomeService,
+    @Inject(MyUnsubscribeService) myUnsubscribeService: MyUnsubscribeService
+  ) {
+-    super(mySubscribeService, myTrackingService, myRemoteHomeService)
++    super(injector)
+  }
+  
+  unsubscribe() {
+    this.myUnsubscribeService.trigger();
+  }
+}
+```
+
+resulting in this much more compact version of a constructor.
+
+```typescript
+@Component(/*...*/)
+export class UnsubscribeToolComponent extends MyAbstractBaseComponent {
+  constructor(
+    @Inject(Injector) injector: Injector
+    @Inject(MyUnsubscribeService) myUnsubscribeService: MyUnsubscribeService
+  ) {
+    super(injector)
+  }
+  
+  /*...*/
+}
+```
+
+## Conclusion
+Handling breaking changes is an act of empathy 💛. You want to protect your users from struggling with your changes. The "inject the injector" pattern we have introduced here helped us a lot and made it very easy to extend our base class without breaking things anymore.
+
+It's always advisable to be more specific instead of using generalized concepts like described in this blog post. Your code will be less abstract, better readable and maintainable. Use this approach only if you have very good reasons to do so 👍 This is a highly specific pattern for our distributed library project. In case you're developing individual components that are not distributed to other developers you most probably should not care for breaking changes and patterns like that.
```

Comment:
"It's always advisable to be more specific instead of using generalized concepts like described in this blog post." - I'm not sure I understand what you mean to say by that. Is it synonymous to 
"It's always advisable to be as specific as possible rather than using generalized concepts like the one we describe in this blog post"?
The last sentence in this paragraph suggests that this is supposed to be disclaimer of sorts. I think I get what you're saying, but maybe something along the lines of
"In case you're developing individual components that are not distributed to other developers you most probably have no need to prevent the occurrence of breaking changes by means of patterns like that"
would be a better choice of words? What developer in their right mind would not care for all sorts of patterns and breaking changes? I think what we want to say is that unless you find yourself working with a large codebase and distributed teams in a typical enterprise organization, dealing with breaking changes is much easier and introducing a pattern like this might not be necessary or even beneficial.

Reply:
You understood me correctly. Actually I wrote this as a word of warning at the beginning of the blog post thinking of the people taking such a pattern and over complicating their simple project. I wanted to clarify that this is only suitable for distributed libraries. I like your last summary and here my updated sentence:


> Introducing a pattern like this might not be necessary or even beneficial if you are not working with a large codebase and distributed teams in a typical enterprise organization. Never use this pattern only to save a few lines of code, you will make your code less understandable.